P.C. GCE cleanup fix to upload log on errors

m-dati opened the pull request:
In recent P.C. GCE preprare_instance issue (poo#161384) the post fail hook `cleanup` routine didn't provided the serial/console log instance_serial.txt because the instance-id resulted empty, no more available at that exit point, and get-serial-port-output command failed.

Here that `id` is retrieved in a different way, used where expected link was empty.

In wait-ssh a timeout parameter added to decrease the default value, for debugging purpose, but it could still be useful in future needs.

See in test Logs available the file "<modulename>-instance_serial.txt" log

- Related ticket: https://progress.opensuse.org/issues/161732
- Needles: N/A
- Verification run: 
  -  https://openqa.suse.de/tests/14523157#step/prepare_instance/90
     FAIL due to GCE issue, 
     but the **instance_id is present** on error too and now `log present`
  -  https://openqa.suse.de/tests/14523287
     EC2 not affected by GCE issue, `log present` ok, failed for other cause: repos download not ready yet a.t.m.

ricardobranco777 reviewed (approved):
LGTM!

asmorodskyi commented on lib/publiccloud/provider.pm:
```diff
@@ -345,6 +345,7 @@ sub create_instances {
     my $url = get_var('PUBLIC_CLOUD_PERF_DB_URI', 'http://larry.qe.suse.de:8086');
 
     foreach my $instance (@vms) {
+        push(@{$self->{provider_client}->{instance_id}}, $instance->{instance_id});
```
```suggestion
        push(@{$self->{my_instance}}, $instance->{instance_id});
```

m-dati replied:
see last version

m-dati commented:
I will update as discussed the code and run more VRs.

m-dati commented:
Verification runs that on final `cleanup`, as when failed as when passed, verify that test log instance_serial.txt produced for existing VM instance:

a) Failure expected, Forced by setting wait timeout 2 sec:

- sle-15-SP5-GCE-Updates-aarch64-Build20240609-1-publiccloud_consoletests @64bit
  https://openqa.suse.de/tests/14566610

- sle-micro-5.4-GCE-BYOS-Updates-x86_64-Build20240609-1-slem_basic  @gce_n1_standard_2
  https://openqa.suse.de/tests/14566613

b) PASS expected from normal end, with logs ok:

- sle-micro-5.4-GCE-BYOS-Updates-x86_64-Build20240609-1-slem_basic @gce_n1_standard_2
  https://openqa.suse.de/tests/14566608

ricardobranco777 commented:
Still Draft / WIP?

pdostal commented on lib/publiccloud/gce.pm:
```diff
@@ -170,7 +171,8 @@ sub cleanup {
     my $region = $self->{provider_client}->{region};
     my $project = $self->{provider_client}->{project_id};
-    my $instance_id = $args->{my_instance}->{instance_id};
+    my $instance_id = ($args->{my_instance}->{instance_id} || $self->{my_instance}->{instance_id}[-1]);
```
Hello Maurizio, I suggest you to do [what I did in ec2](https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/115e3ac750c822eece081478443effe48e1a8593/lib/publiccloud/ec2.pm#L213). It simplifies the code quite a lot.

m-dati replied:
Interesting and smart solution, I'd like more clarifications on it, but my idea is to preserve the `create_instances` possible **multiple instances** creation, for which I store each instance_id into that array, then `cleanup` pop out the last one [-1] that, when routine is triggered on-error, it is the guilty id where error occurred.

m-dati commented:
After discussed other solutions, I just pushed an update to test VRs, before to define final code.

asmorodskyi commented:
nice ! I like new version !

m-dati commented:
All C.I. in last commit failing, even having no relevant change compared with before passed.

ricardobranco777 commented on lib/publiccloud/provider.pm:
```diff
@@ -705,6 +705,23 @@ sub terraform_param_tags
+=cut
+
+sub terraform_query {
+    my ($self, $qry) = @_;
```
Why not use full variable name?  We're just saving 2 characters here.

```suggestion
    my ($self, $query) = @_;
```

m-dati replied:
done

mimi1vx commented on lib/publiccloud/provider.pm:
```diff
@@ -705,6 +705,23 @@ sub terraform_param_tags
+    # default query
+    $qry //= ".vm_name.value[0]";
+    assert_script_run('cd ~/terraform');
+    $qry = script_output("terraform output -json | jq -r \"$qry\" 2>/dev/null", proceed_on_failure => 1);
```
I don't think reusing `query` for return value is good idea 
    
```suggestion
    my $ret = script_output("terraform output -json | jq -r \"$qry\" 2>/dev/null", proceed_on_failure => 1);
```

m-dati replied:
done

pdostal commented:
I'm sorry but I don't understand what "terraform query" means and why such "query" should have default of `.vm_name.value[0]`.

We call terraform is several places which we should probably replace by this new subroutine. Our use cases are:
 * `my $output = decode_json(script_output("terraform output -json"));`
 * `my $instance_id = script_output('terraform output -json | jq -r ".vm_name.value[0]"', proceed_on_failure => 1);`
 * `--extra-vars \"\$(terraform output --json | jq 'with_entries(.value |= .value)')\"", timeout => 300);`

I also don't understand why `publiccloud::provider::terraform_query` isn't in `publiccloud::utils::terraform_query` and why it needs `$self`.

What was wrong on my yesterday's suggestion?
```perl
my $instance_id = script_output('terraform output -json | jq -r ".vm_name.value[0]"', proceed_on_failure => 1);
record_info('WARN", "Didn't detect instance ID") if ($instance_id eg "");
```

Thank you

m-dati commented:
Hi Pavel, sorry but it's me, kindly, not understanding your question on a so clear context: 
Please what's your problem with 'default', can you elaborate?
What's wrong with having a routine that, as you mentioned, possibly could be used in more places? 
For Sure this routine can be better tuned,moved wherever is better, fixed and improved and suggestions welcome.

asmorodskyi commented:
> Good morning Maurizio,
> 
> I'm sorry but I don't understand what "terraform query" means and why such "query" should have default of `.vm_name.value[0]`.
> 
> We call `terraform output` is several places which we should probably replace by this new subroutine. Our use cases are:
> 
> * `my $output = decode_json(script_output("terraform output -json"));`
> * `my $instance_id = script_output('terraform output -json | jq -r ".vm_name.value[0]"', proceed_on_failure => 1);`
> * `--extra-vars \"\$(terraform output --json | jq 'with_entries(.value |= .value)')\"", timeout => 300);`
> 
> I also don't understand why `publiccloud::provider::terraform_query` isn't in `publiccloud::utils::terraform_query` and why it needs `$self`.
> 
> What was wrong on my yesterday's suggestion?
> 
> ```perl
> my $instance_id = script_output('terraform output -json | jq -r ".vm_name.value[0]"', proceed_on_failure => 1);
> record_info('WARN", "Didn't detect instance ID") if ($instance_id eg "");
> ```
> 
> Thank you

1. I agree that `terraform_query` is confusing name for what function is doing , but I don't think that we need go further and try to create function content which would fit the name . Instead we better just rename a function to something which represents better the content . What about `get_vm_name_from_terraform` ?
2. I don't understand your motivation to move this function to utils . We have everything else related to terraform in provider ( e.g. `terraform_apply`)  so for me it makes perfect sense to keep this one together with them

pdostal commented:
> 1. I agree that `terraform_query` is confusing name for what function is doing , but I don't think that we need go further and try to create function content which would fit the name . Instead we better just rename a function to something which represents better the content . What about `get_vm_name_from_terraform` ?

We may create `get_vm_name_from_terraform` but then such subroutine is very specific and does not need any parameters. Such subroutine will probably also contain conditions for each provider because terraform variable `vm_name`  returns `aws_instance.openqa.*.id` for ec2 provider but `google_compute_instance.openqa.*.name` for gce provider and I am worried that for GCE we need terraform to output `google_compute_instance.openqa.*.instance_id`. So such simple subroutine would increase complexity.

> 2. I don't understand your motivation to move this function to utils . We have everything else related to terraform in provider ( e.g. `terraform_apply`)  so for me it makes perfect sense to keep this one together with them

Because subroutine in `provider` object requires `$self->{provider}` to be available in `post_fail_hook` in every test module. I tried to descreese the complexity by getting rid of this dependency.

pdostal commented:
```
$ terraform output --help
Usage: terraform [global options] output [options] [NAME]

  Reads an output variable from a Terraform state file and prints
  the value. With no additional arguments, output will display all
  the outputs for the root module.  If NAME is not specified, all
  outputs are printed.
```

So we are not querying Terraform, we are instructing it to output variable. And we specify that variable by its NAME.

And I think this is important because it makes the code readable and understandable.

mloviska commented on lib/publiccloud/instance.pm:
```diff
@@ -372,7 +372,7 @@ Return:
     my ($self, %args) = @_;
     # Input parameters, see description in above head2 - Parameters section:
-    $args{timeout} //= 600;
+    $args{timeout} //= get_var('PUBLIC_CLOUD_SSH_TIMEOUT', 600);
```
you can remove `//` operator as get_var returns default now

grisu48 replied:
But it would only use `get_var` if not explicitly defined in `%args`, so I think removing the `//` would alter the intended use case.

m-dati replied:
As we discussed, this  expression was introduced here just to tweak the _default_ value for `debugging purposes`, via test settings, _when the input parameter is empty_, either to force a timeout (just assigning in clone command a proper low value, like PUBLIC_CLOUD_SSH_TIMEOUT=1 or 2,3..) or to force never timeout (a very high value). The `normal` routine flow shall still rely in the last coded default. But indeed do **never apply** it when a `timeout` value is passed in the routine call: it that last case that value shall be used.

m-dati replied:
As explained, I realized that just when the input value is explicitly defined, the debug-tweaking cannot be used anymore.
Therefore I changed that parameter to be prior when set for debuugging, and the `input // constant` when empty, that is the normal code-flow.

pdostal commented on lib/publiccloud/provider.pm:
```diff
@@ -705,6 +705,23 @@ sub terraform_param_tags
+
+sub terraform_query {
+    my ($self, $qry) = @_;
+    # default query
+    $qry //= ".vm_name.value[0]";
```
```suggestion
```

asmorodskyi reviewed (approved):
LGTM

m-dati commented on lib/publiccloud/gce.pm:
```diff
@@ -170,7 +171,8 @@ sub cleanup {
     my $region = $self->{provider_client}->{region};
     my $project = $self->{provider_client}->{project_id};
-    my $instance_id = $args->{my_instance}->{instance_id};
+    my $instance_id = ($args->{my_instance}->{instance_id} || $self->{provider_client}->{instance_id}[-1]);
```
see last version

m-dati commented:
VR rerun:
- FAIL (forced adding PUBLIC_CLOUD_SSH_TIMEOUT=2): 
  https://openqa.suse.de/tests/14580003#step/prepare_instance/95
  https://openqa.suse.de/tests/14580003/logfile?filename=prepare_instance-instance_serial.txt

- PASS (forced timeout removed):
  https://openqa.suse.de/tests/14580009

ricardobranco777 commented on lib/publiccloud/provider.pm:
```diff
@@ -705,6 +705,25 @@ sub terraform_param_tags
+sub get_terraform_output {
+    my ($self, $jq_query) = @_;
+    assert_script_run('cd ~/terraform');
+    my $res = script_output("terraform output -json | jq -r \"$jq_query\" 2>/dev/null", proceed_on_failure => 1);
```
I would use single-quotes here because JSON queries can use double quotes...

```suggestion
    my $res = script_output("terraform output -json | jq -r '$jq_query' 2>/dev/null", proceed_on_failure => 1);
```

m-dati replied:
ok, but my concern was that single quotes prevented the inner variable expansion

mimi1vx replied:
+ Single and double quotes have different meanings in bash ( same as Perl fun )

m-dati replied:
yes, but at that level the perl syntax comes first, so question: will that perl variable be safe with single quotes?

mimi1vx replied:
if they are inside double-quoted string they should be taken as normal characters so Perl var should correctly eval

```Perl
$ use Modern::Perl
$ my $test = "testuji string"
testuji string
$ say "test - '$test'"
test - 'testuji string'
1
$ say 'test - "$test"'
test - "$test"
```

m-dati replied:
ok :smile:

m-dati replied:
But I have different results in my oneliner:
```
perl -e '
my $q = "This is a test STRING.";
print "Testing: \"$q\"";
print "\n";
'
```
`Testing: "This is a test STRING."`
```
perl -e '
my $q = "This is a test STRING.";
print "Testing: '$q'";
print "\n";
'
```
`Testing:`

but even failure with:
```
perl -e '
my $q = "This is a test STRING.";
print 'Testing: "$q" ';
print "\n";
'
```
`syntax error at -e line 3, near "Testing:"
Execution of -e aborted due to compilation errors.`

Therefore I should run on OSD that change porposal, if really it works.

perlpunk replied:
@m-dati that's because the `'` is ending your perl input and then the shell tries to resolve the environment variable `$q`.
Try this:
```
perl -e '
my $q = "This is a test STRING.";
print "Testing: '$USER'";
print "\n";
'
```
Also, as a general note, always use `-w` in oneliners. I have seen so many questions over the years where the answer was simple "enable warnings and then you see yourself why it doesn't work". (not in this case though)
Also, use `-E` in oneliners, so you can just use `say` and save the `"\n"` thing.
e.g.
```
perl -wE '
my $q = "This is a test STRING.";
say "Testing: '$USER'";
'
```

m-dati replied:
Ok, in the editor I have:
```
...
print "Testing1: ==>'$q' --> $q ";
```
`Testing1: ==>'This is a test STRING.' --> This is a test STRING.`

Updating.

m-dati replied:
@perlpunk sorry I late reading your reply: yes as I wrote, I noted the external quote impact :slightly_smiling_face:. But indeed your suggestions on options are really useful. Thank you!

m-dati commented:
Now checking the VRs then will remove WIP

m-dati commented:
VR rerun after last update:

- to FAIL: https://openqa.suse.de/tests/14581672

- to PASS: https://openqa.suse.de/tests/14581675

mpagot commented on lib/publiccloud/instance.pm:
```diff
@@ -372,7 +372,7 @@ Return:
     my ($self, %args) = @_;
     # Input parameters, see description in above head2 - Parameters section:
-    $args{timeout} //= 600;
+    $args{timeout} = get_var('PUBLIC_CLOUD_SSH_TIMEOUT', $args{timeout} // 600);
```
```suggestion
    $args{timeout} //= get_var('PUBLIC_CLOUD_SSH_TIMEOUT', 600);
```

m-dati replied:
No Mike sorry :smile: I just changed that code that was quite as you're proposing now, see the thread ...

m-dati replied:
I hope that link will clarify the last change

mpagot commented on lib/publiccloud/provider.pm:
```diff
@@ -705,6 +705,25 @@ sub terraform_param_tags
+    my ($self, $jq_query) = @_;
+    assert_script_run('cd ~/terraform');
+    my $res = script_output("terraform output -json | jq -r '$jq_query' 2>/dev/null", proceed_on_failure => 1);
+    script_run('cd');
```
why this last `cd` is needed?

m-dati replied:
Because of previous cd ..., this last go back to previous place.

mpagot commented on lib/publiccloud/instance.pm:
Maybe change in this file could have UT in ` t/13_publiccloud.t`

grisu48 replied:
I suggest to proceed for now without unit test.

m-dati commented on lib/publiccloud/provider.pm:
```diff
@@ -705,6 +705,25 @@ sub terraform_param_tags
+
+sub get_terraform_output {
+    my ($self, $jq_query) = @_;
+    assert_script_run('cd ~/terraform');
```
If SAP will still need this publiccloud code-flow, ok I will try to align it

mpagot replied:
Applying this customization open the door to SLES4SAP to use it. 
It could be useful in https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/42d9cf281e7a021a04d9583293ad75f8355cea1a/lib/sles4sap_publiccloud.pm#L726

This function at the moment is implemented reading the Ansible inventory but I think terraform output has the same info too.

All in all, hardcoded path aside, this function is doing a real generic task: 
- run `terraform output`
- pipe the output through `jq` with a jq query string provided by the caller
- return the result as string

Anyway totally up to you

m-dati replied:
Well, if SAP does **not using** anywhere the `gce->cleanup` routine here updated, the update proposal could eventually part of a new dedicated PR, if ok to you.

But if SAP does use it, change needed now to align the usage.

grisu48 replied:
@mpagot Does SAP already have a setting for this or how would you suggest to have this done?

mpagot replied:
> SAP does not using anywhere the gce->cleanup 
This is true, no any SLES4SAP use that cleanup function.

> Does SAP already have a setting for this or how would you suggest to have this done?
This PR as it is now does not require any change in any existing SAP test
My comment was just an abstract and generic idea looking at the code of this particular function, out of any context, not considering the `cleanup()` caller or anything. Looking at the code all parts are looking generic with no dependency on any QE-C or QE-SAP specific deployment strategy and folder organization. From some point of view it can be described as "a perl function calling `terraform output` and extracting some string from its output using a jq query provided by the caller"
The only not generic part in it is the `cd ~/teraform`. Generally speaking it is not generic, it suppose that the terraform state files are exactly and only in a specific folder.

This PR can be merged as it is right now from my point of view.

m-dati replied:
Thank you much Michele for your clarification. Then we proceed with the current status.

Eventually, after merged, we can deal with improvements to the `get` routine, like on-demand customized work-directory, useful for SAP too.

mpagot replied:
Something to check about the terraform path https://openqa.suse.de/tests/14598491#step/ssh_interactive_end/32 (does not means this PR is wrong)

m-dati commented:
Hi all, given this [comment](https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/19453#discussion_r1636677483), would it make sense to save/restore the calling directory place in `get_terraform_output` (add some complexity)?

mpagot reviewed (approved):
LGTM

m-dati commented:
After last clarifications in https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/19453#discussion_r1637856750, we can proceed with the current version merge. 
Many thanks to all the reviewers.
